types/react-query/index.d.ts

```diff
@@ -1,142 +1,370 @@
-// Type definitions for react-query 0.3
+// Type definitions for react-query 1.1
 // Project: https://github.com/tannerlinsley/react-query
 // Definitions by: Lukasz Fiszer <https://github.com/lukaszfiszer>
 //                 Jace Hensley <https://github.com/jacehensley>
 //                 Matteo Frana <https://github.com/matteofrana>
+//                 Igor Oleinikov <https://github.com/igorbek>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Minimum TypeScript Version: 3.7
 
-import { ComponentType } from 'react';
+import * as React from 'react';
+import * as _ from 'ts-toolbelt';
 
-// overloaded useQuery function with pagination
-export function useQuery<TResult, TVariables extends object>(
-    queryKey: QueryKey<TVariables>,
-    queryFn: QueryFunction<TResult, TVariables>,
-    options: QueryOptionsPaginated<TResult>
-): QueryResultPaginated<TResult, TVariables>;
+// overloaded useQuery function
+export function useQuery<TResult, TKey extends AnyQueryKey>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    queryFn: QueryFunction<TResult, TKey>,
+    config?: QueryOptions<TResult>,
+): QueryResult<TResult>;
 
-export function useQuery<TResult, TVariables extends object>(
-    queryKey: QueryKey<TVariables>,
-    queryFn: QueryFunction<TResult, TVariables>,
-    options?: QueryOptions<TResult>
-): QueryResult<TResult, TVariables>;
+export function useQuery<TResult, TKey extends string>(
```

Comment:
Did you considered merging those 2 overloads into one? The difference in `TKey` arg could be handled by wrapping it with a conditional type before passing into `QueryFunction`

Reply:
I'm trying to not overuse complex type constructs such as conditional types if you can achieve the same with overloading or any other simpler constructs. TS usually works better with them.
Also, I wanted to be sure that what a key means is well defined. In this case, the key is a tuple with serializable elements.
In this overload, I probable needed to call this type `TSingleKey` or something like that.